cpp/arcticdb/storage/test/test_storage_exceptions.cpp

```diff
@@ -0,0 +1,163 @@
+protected:
+    std::unique_ptr<acs::Storage> storage;
+
+    void SetUp() override {
+        if (!fs::exists(TEST_DATABASES_PATH)) {
+            fs::create_directories(TEST_DATABASES_PATH);
+        }
+
+        storage = GetParam()->create();
+    }
+
+    void TearDown() override {
+        storage.reset();
+
+        if (fs::exists(TEST_DATABASES_PATH)) {
+            fs::remove_all(TEST_DATABASES_PATH);
+        }
+    }
+};
+
+TEST_P(GenericStorageTest, WriteDuplicateKeyException) {
+    ac::entity::AtomKey k = ac::entity::atom_key_builder().gen_id(0).build<ac::entity::KeyType::VERSION>("sym");
+
+    acs::KeySegmentPair kv(k);
+    kv.segment().set_buffer(std::make_shared<ac::Buffer>());
+
+    storage->write(std::move(kv));
+
+    ASSERT_TRUE(storage->key_exists(k));
+
+    acs::KeySegmentPair kv1(k);
+    kv1.segment().set_buffer(std::make_shared<ac::Buffer>());
+
+    ASSERT_THROW({
+        storage->write(std::move(kv1));
+    },  acs::DuplicateKeyException);
+
+}
+
+TEST_P(GenericStorageTest, ReadKeyNotFoundException) {
+    ac::entity::AtomKey k = ac::entity::atom_key_builder().gen_id(0).build<ac::entity::KeyType::VERSION>("sym");
+
+    ASSERT_TRUE(!storage->key_exists(k));
+    ASSERT_THROW({
+        storage->read(k, acs::ReadKeyOpts{});
+    },  acs::KeyNotFoundException);
+
+}
+
+TEST_P(GenericStorageTest, UpdateKeyNotFoundException) {
+    ac::entity::AtomKey k = ac::entity::atom_key_builder().gen_id(0).build<ac::entity::KeyType::VERSION>("sym");
+
+    acs::KeySegmentPair kv(k);
+    kv.segment().header().set_start_ts(1234);
+    kv.segment().set_buffer(std::make_shared<ac::Buffer>());
+
+    ASSERT_TRUE(!storage->key_exists(k));
+    ASSERT_THROW({
+        storage->update(std::move(kv), acs::UpdateOpts{});
+    },  acs::KeyNotFoundException);
+
+}
+
+TEST_P(GenericStorageTest, RemoveKeyNotFoundException) {
+    ac::entity::AtomKey k = ac::entity::atom_key_builder().gen_id(0).build<ac::entity::KeyType::VERSION>("sym");
+
+    ASSERT_TRUE(!storage->key_exists(k));
+    ASSERT_THROW({
+        storage->remove(k, acs::RemoveOpts{});
+    },  acs::KeyNotFoundException);
+
+}
+
+INSTANTIATE_TEST_SUITE_P(
+        AllStoragesCommonTests,
+        GenericStorageTest,
+        ::testing::Values(std::make_shared<LMDBStorageFactory>())
+);
+
+// LMDB Storage specific tests
+
+class LMDBStorageTestBase : public ::testing::Test {
+protected:
+    void SetUp() override {
+        if (!fs::exists(TEST_DATABASES_PATH)) {
+            fs::create_directories(TEST_DATABASES_PATH);
+        }
+    }
+    void TearDown() override {
+        if (fs::exists(TEST_DATABASES_PATH)) {
+            fs::remove_all(TEST_DATABASES_PATH);
+        }
+    }
+};
+
+TEST_F(LMDBStorageTestBase, WriteMapFullError) {
+    LMDBStorageFactory factory(32ULL * (1ULL << 10));
+    auto storage = std::unique_ptr<acs::lmdb::LmdbStorage>(dynamic_cast<acs::lmdb::LmdbStorage*>(factory.create().release()));
```

Comment:
?? You can just call `factory.create()` right? What's the point of all this type manipulation?